api/agent/agent.go

```diff
@@ -867,22 +867,33 @@ func (a *agent) runHot(ctx context.Context, caller slotCaller, call *call, tok R
 		return
 	}
 
-	needsPull, err := cookie.ValidateImage(ctx)
+	err = cookie.AuthImage(ctx)
 	if tryQueueErr(err, errQueue) != nil {
 		return
 	}
 
+	needsPull, err := cookie.ValidateImage(ctx)
 	if needsPull {
-		ctx, cancel := context.WithTimeout(ctx, a.cfg.HotPullTimeout)
-		err = cookie.PullImage(ctx)
-		cancel()
-		if ctx.Err() == context.DeadlineExceeded {
-			err = models.ErrDockerPullTimeout
+		// pull image ctx scope with HotPullTimeout
+		{
```

Comment:
do we need the extra brackets?

Reply:
yes, it's for the new 'ctx' in that scope with specific timeout for the pull.

Reply:
subsequent ValidateImage() uses outer scope ctx while PullImage() uses inner ctx with hot pull timeout.

Reply:
right, which is hard to read, I would use something like `pullCtx` to make this behavior explicit, but it is a super nit... 😈